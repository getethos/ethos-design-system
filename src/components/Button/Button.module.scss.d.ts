// This file is automatically generated.
// Please do not change this file!
interface CssExports {
  'Black': string;
  'BlackOutline': string;
  'Button': string;
  'Medium': string;
  'Small': string;
  'Stateful': string;
  'Unstyled': string;
  'White': string;
  'WhiteCTA': string;
  'WhiteOutline': string;
  'arrowIcon': string;
  'arrowIconInline': string;
  'backArrowIconInline': string;
  'fullWidth': string;
  'isSelected': string;
}
export const cssExports: CssExports;
export default cssExports;
