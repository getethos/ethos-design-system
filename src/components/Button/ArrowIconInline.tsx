import * as React from 'react'
import styles from './Button.module.scss'
export const ArrowIconInline = ({ shouldFlip }: { shouldFlip?: boolean }) => {
  return (
    <svg
      width="14"
      height="14"
      viewBox="0 0 14 14"
      fill="none"
      xmlns="http://www.w3.org/2000/svg"
      className={
        shouldFlip ? styles.backArrowIconInline : styles.arrowIconInline
      }
    >
      <path d="M7.00016 0.333374L5.82516 1.50837L10.4752 6.16671H0.333496V7.83337H10.4752L5.82516 12.4917L7.00016 13.6667L13.6668 7.00004L7.00016 0.333374Z" />
    </svg>
  )
}
