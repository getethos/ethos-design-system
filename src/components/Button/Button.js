import React from 'react'
import PropTypes from 'prop-types'
import useIncludes from '../../hooks/useIncludes.js'
import useInvalid from '../../hooks/useInvalid.js'
import useDebounce from '../../hooks/useDebounce.js'

import styles from './Button.module.scss'

const DEFAULT_DEBOUNCE_DURATION_MS = 0

/* @getethos/design-system/Button.js

   Legend:

   - `PrivateButton` is a private component that returns an element with CSS classes.
   - `Buttons.scss` (not imported here) implements those CSS classes.
   - `ButtonFactory` is a HOC that creates public components with correct props.
   - `<Button.Tiny.Blue>`, etc. are the set of Design-approved public buttons.
   ========================================================================== */

/**
 * `PrivateButton` is a private component that returns a <button> with CSS classes and
 * other HTML attributes like `disabled` and `type`.
 *
 * Note that it is ignorant of which combinations Design considers legal or
 * illegal. Legal combinations must be declared via the enumerated exports at
 * the end of this file.
 *
 * The public props (Button.PUBLIC_PROPS) may be passed in downstream.
 * Other props may only be specified in this file.
 *
 * `type` is important to understand because it determines whether the button
 * is a regular button or a form submission button (<button type="submit">).
 * https://developer.mozilla.org/en/docs/Web/HTML/Element/button
 *
 * @param  {object}   props             props
 * @param  {String}   props.children    The button text to display
 * @param  {Boolean}  props.disabled    Whether the button is disabled
 * @param  {String}   props.type        <button type="button|submit">
 * @param  {String}   props.size        The size of the button.
 * @param  {String}   props.style       The color style of the button.
 * @param  {Boolean}  props.arrowIcon   Whether the arrow icon is displayed
 *                                      (behaves differently when fullWidth)
 * @param  {Boolean}  props.backArrowIcon   Back arrow icon
 * @param  {Boolean}  props.refreshIcon   refresh icon
 * @param  {number}   props.debounceDurationMs   Length of time to debounce button
 *                                               (if 0, no debounce)
 */

function PrivateButton({
  children,
  disabled,
  fullWidth,
  role,
  ariaLabelId,
  name,
  type,
  size,
  style,
  onClick,
  isSelected, // only used for Button.*.Stateful, aka SelectableHtmlButtonGroup
  arrowIcon,
  backArrowIcon,
  refreshIcon,
  debounceDurationMs = DEFAULT_DEBOUNCE_DURATION_MS,
  ...rest
}) {
  const [isDebounced, debouncedOnClick] = useDebounce(
    onClick,
    debounceDurationMs
  )

  // Verify that size, type, and style were valid enum values
  const [isLegalSize] = useIncludes(PrivateButton.SIZES)
  isLegalSize(size)
  const [isLegalType] = useIncludes(PrivateButton.HTML_TYPES)
  isLegalType(type)
  const [isLegalStyle] = useIncludes(PrivateButton.STYLES)
  isLegalStyle(style)

  // Verify that other required props were supplied
  const [, includesKeysOrThrow] = useIncludes(['data-tid'])
  includesKeysOrThrow(rest)

  // Verify that no invalid props were supplied
  const [includesInvalid] = useInvalid(Object.keys(PrivateButton.PUBLIC_PROPS))
  includesInvalid(rest)

  // Generate list of css classes
  // style may be 'Stateful White' so we need to split that
  const classNames = ['Button', size].concat(style.split(' '))
  if (fullWidth) classNames.push('fullWidth')
  if (isSelected) classNames.push('isSelected')
  if (arrowIcon || backArrowIcon) classNames.push('arrowIcon')
  if (refreshIcon) classNames.push('refreshIcon')

  let checked
  if (isSelected) {
    classNames.push('isSelected')
    checked = { 'aria-checked': isSelected }
  }

  const cssModulesClasses = classNames.reduce(
    (accumulator, klass, i) =>
      accumulator.concat(
        `${styles[klass]}${i < classNames.length - 1 ? ' ' : ''}`
      ),
    ''
  )

  return (
    <button
      {...checked}
      className={cssModulesClasses}
      disabled={disabled || isDebounced}
      type={type}
      name={name}
      onClick={debouncedOnClick}
      data-tid={rest['data-tid']}
      role={role}
    >
      {backArrowIcon && ArrowIconInline(true)}
      {ariaLabelId ? <span id={ariaLabelId}>{children}</span> : children}
      {arrowIcon && ArrowIconInline()}
      {refreshIcon && <RefreshIcon />}
    </button>
  )
}

const ArrowIconInline = (shouldFlip) => {
  return (
    <svg
      width="14"
      height="14"
      viewBox="0 0 14 14"
      fill="none"
      xmlns="http://www.w3.org/2000/svg"
      className={
        shouldFlip ? styles.backArrowIconInline : styles.arrowIconInline
      }
    >
      <path d="M7.00016 0.333374L5.82516 1.50837L10.4752 6.16671H0.333496V7.83337H10.4752L5.82516 12.4917L7.00016 13.6667L13.6668 7.00004L7.00016 0.333374Z" />
    </svg>
  )
}

function RefreshIcon() {
  return (
    <svg
      width="24"
      height="24"
      viewBox="0 1 24 24"
      fill="none"
      xmlns="http://www.w3.org/2000/svg"
      className={styles.refreshIconInner}
    >
      <g clipPath="url(#clip0)">
        <path
          d="M5.01064 11.9989C5.01128 9.9477 5.78291 7.90328 7.34659 6.3396C10.474 3.21223 15.5314 3.21772 18.6568 6.34313C21.7822 9.46855 21.7877 14.5259 18.6603 17.6533C16.0211 20.2925 12.0175 20.6898 8.94831 18.88L10.42 17.4083C12.6478 18.4757 15.3994 18.0858 17.2461 16.2391C19.5881 13.8971 19.5893 10.0917 17.2487 7.75116C14.9082 5.41063 11.1028 5.41182 8.76081 7.75381C7.58627 8.92834 7.02701 10.4634 7.0336 11.9983L11.5888 11.9969L6.63595 16.9497L1.6862 12L5.01064 11.9989Z"
          fill="white"
        />
      </g>
      <defs>
        <clipPath id="clip0">
          <rect width="24" height="24" fill="white" />
        </clipPath>
      </defs>
    </svg>
  )
}

PrivateButton.HTML_TYPES = { BUTTON: 'button', SUBMIT: 'submit' } // read the docs^

PrivateButton.PUBLIC_PROPS = {
  ariaLabelId: PropTypes.string,
  role: PropTypes.string,
  children: PropTypes.node,
  'data-tid': PropTypes.string,
  disabled: PropTypes.bool,
  fullWidth: PropTypes.bool,
  name: PropTypes.string,
  onClick: PropTypes.func,
  isSelected: PropTypes.bool, // only for Stateful style
  type: PropTypes.oneOf(Object.values(PrivateButton.HTML_TYPES)),
  arrowIcon: PropTypes.bool,
  backArrowIcon: PropTypes.bool,
  debounceDurationMs: PropTypes.number,
}

PrivateButton.SIZES = {
  // ***** ONLY USE SMALL IN UniversalNavbar! ***** //
  SMALL: 'Small',
  MEDIUM: 'Medium',
  UNSIZED: 'Unsized',
}

// a space in the value will generate two different classes.
PrivateButton.STYLES = {
  BLACK: 'Black',
  BLACK_OUTLINE: 'BlackOutline',
  SALAMANDER: 'Salamander',
  CYPRESS: 'Cypress',
  WHITE_OUTLINE: 'WhiteOutline',
  STATEFUL: 'Stateful',
  STATEFUL_WHITE: 'Stateful White',
  STATEFUL_FLOATING: 'Stateful Floating',

  // For semantic <buttons> that are not styled as buttons:
  UNSTYLED: 'Unstyled',

  // Only used for CMS 'Check my price' CTA button on hero
  WHITE_CTA: 'WhiteCTA',
}

PrivateButton.propTypes = {
  ...PrivateButton.PUBLIC_PROPS,
  size: PropTypes.oneOf(Object.values(PrivateButton.SIZES)),
  style: PropTypes.oneOf(Object.values(PrivateButton.STYLES)),
}

PrivateButton.defaultProps = { type: PrivateButton.HTML_TYPES.BUTTON }

function ButtonFactory(privateProps) {
  const PublicButtonComponent = (downstreamProps) => {
    return <PrivateButton {...downstreamProps} {...privateProps} />
  }

  return PublicButtonComponent
}

export const Button = {
  Medium: {
    Black: ButtonFactory({
      size: PrivateButton.SIZES.MEDIUM,
      style: PrivateButton.STYLES.BLACK,
    }),
    BlackOutline: ButtonFactory({
      size: PrivateButton.SIZES.MEDIUM,
      style: PrivateButton.STYLES.BLACK_OUTLINE,
    }),
    Salamander: ButtonFactory({
      size: PrivateButton.SIZES.MEDIUM,
      style: PrivateButton.STYLES.SALAMANDER,
    }),
    Cypress: ButtonFactory({
      size: PrivateButton.SIZES.MEDIUM,
      style: PrivateButton.STYLES.CYPRESS,
    }),
    WhiteOutline: ButtonFactory({
      size: PrivateButton.SIZES.MEDIUM,
      style: PrivateButton.STYLES.WHITE_OUTLINE,
    }),
    Stateful: {
      Default: ButtonFactory({
        size: PrivateButton.SIZES.MEDIUM,
        style: PrivateButton.STYLES.STATEFUL,
      }),
      White: ButtonFactory({
        size: PrivateButton.SIZES.MEDIUM,
        style: PrivateButton.STYLES.STATEFUL_WHITE,
      }),
      Floating: ButtonFactory({
        size: PrivateButton.SIZES.MEDIUM,
        style: PrivateButton.STYLES.STATEFUL_FLOATING,
      }),
    },
  },
  // ***** ONLY USE SMALL IN UniversalNavbar! ***** //
  Small: {
    BlackOutline: ButtonFactory({
      size: PrivateButton.SIZES.SMALL,
      style: PrivateButton.STYLES.BLACK_OUTLINE,
    }),
    Black: ButtonFactory({
      size: PrivateButton.SIZES.SMALL,
      style: PrivateButton.STYLES.BLACK,
    }),
  },
  Unstyled: ButtonFactory({
    size: PrivateButton.SIZES.UNSIZED,
    style: PrivateButton.STYLES.UNSTYLED,
  }),
  WhiteCTA: ButtonFactory({
    size: PrivateButton.SIZES.UNSIZED,
    style: PrivateButton.STYLES.WHITE_CTA,
  }),
}
