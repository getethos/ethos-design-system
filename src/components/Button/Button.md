Set of Design-approved public buttons.

```jsx
<Button.Medium.Black>Button.Medium.Black</Button.Medium.Black>
```

```jsx
<Button.Medium.Black fullWidth>
  Button.Medium.Black fullWidth
</Button.Medium.Black>
```

```jsx
<Button.Medium.BlackOutline>
  Button.Medium.BlackOutline
</Button.Medium.BlackOutline>
```

```jsx
<Button.Medium.WhiteOutline>
  Button.Medium.WhiteOutline
</Button.Medium.WhiteOutline>
```

```jsx
<Button.Medium.Stateful.Default>
  Button.Medium.Stateful.Default
</Button.Medium.Stateful.Default>
```

```jsx
<Button.Medium.Stateful.Default isSelected={true}>
  Button.Medium.Stateful.Default
</Button.Medium.Stateful.Default>
```

```jsx
<Button.Medium.Black arrowIcon>
  Button.Medium.Black with arrowIcon
</Button.Medium.Black>
```

```jsx
<Button.Medium.Black fullWidth arrowIcon>
  Button.Medium.Black + fullWidth + arrowIcon
</Button.Medium.Black>
```

```jsx
<Button.Medium.Black backArrowIcon>
  Button.Medium.Black + backArrowIcon
</Button.Medium.Black>
```

```jsx
<Button.Medium.Black fullWidth backArrowIcon>
  Button.Medium.Black + fullWidth + backArrowIcon
</Button.Medium.Black>
```

```jsx
<Button.Unstyled>Button.Unstyled</Button.Unstyled>
```

<strong>The following are only for use in indicated places</strong>

PLEASE ONLY USE IN UniversalNavbar! In fact, have a look at
UniversalNavbar component to see this in use.

```jsx
<Button.Small.BlackOutline>
  Only use for UniversalNavbar!
</Button.Small.BlackOutline>
```

PLEASE ONLY USE FOR CMS hero CTA

```jsx
<Button.WhiteCTA>Button.WhiteCTAâ€”Only for CMS hero CTA</Button.WhiteCTA>
```
