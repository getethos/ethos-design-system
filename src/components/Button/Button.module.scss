// @import "../Animation.scss";
// @import "../Colors.scss";
// @import "../Type/Type.scss";
@import '../Media/Media.scss';

// This is currently a one-off implementation for UniversalNavbar
@import './Button.small.shame.scss';

/* @getethos/design-system/Button.scss
   ========================================================================== */

// Defaults

:root {
  --ButtonMediumStatefulMinWidth: 156px;
}

button,
.Button {
  border: 1px solid transparent;
  font-family: var(--Theinhardt-font-stack);
  color: var(--GrayPrimary--translucent);

  transition: var(--transition-default);
  transition-property: background-color, border-color, color;

  &:not([disabled]) {
    cursor: pointer;
  }
}

// Dimensions

.Button.Medium {
  $font-size: 19;
  $line-height: 24;
  font-size: $font-size * 1px;
  line-height: $line-height / $font-size;
  padding: 17px 16px 13px;
}

.Button.fullWidth {
  width: 100%;
}

.arrowIconInline {
  margin-right: 3px;
  margin-top: 2px;
  margin-left: 7px;
}

.Button.arrowIcon {
  display: flex;
  justify-content: space-between;
}

.Button.arrowIcon.fullWidth {
  @include for-tablet-and-laptop {
    padding-right: 40px;
    padding-left: 40px;
  }
  @include for-desktop-only {
    padding-right: 48px;
    padding-left: 48px;
  }
  @include for-phone-only {
    max-width: 480px;
  }
}
// Styles

@mixin light-on-dark-font-smoothing {
  -webkit-font-smoothing: antialiased; // stackoverflow.com/a/17927764
  -moz-osx-font-smoothing: grayscale;
}

.Button.Unstyled {
  background: none;
  border: none;
  padding: 0;
}

.Button.Black {
  background-color: var(--GrayPrimary--translucent);
  color: var(--White);
  @include light-on-dark-font-smoothing;

  &:not([disabled]) {
    &:hover {
      background-color: var(--GrayDarkHover--translucent);
    }

    &:active {
      background-color: var(--GrayDarkHover--translucent);
    }
  }

  &[disabled] {
    border-color: var(--Transparent--white);
    background-color: var(--GrayStrokeAndDisabled--translucent);
    color: var(--White);
  }
}

.Button.BlackOutline {
  border-color: var(--GrayPrimary--opaque);
  background-color: var(--Transparent--white);
  color: var(--GrayPrimary--opaque);

  &:not([disabled]) {
    &:hover {
      border-color: var(--GrayDarkHover--translucent);
      background-color: var(--GrayDarkHover--translucent);
      color: var(--White);
    }

    &:active {
      border-color: var(--GrayDarkHover--translucent);
      background-color: var(--GrayDarkHover--translucent);
      color: var(--White);
    }
  }

  &[disabled] {
    border-color: var(--Transparent--white);
    background-color: var(--GrayStrokeAndDisabled--translucent);
    color: var(--White);
  }
}

.Button.WhiteOutline {
  border-color: var(--White);
  background-color: var(--Transparent--white);
  color: var(--White);

  &:not([disabled]) {
    &:hover {
      border-color: var(--White);
      background-color: var(--White);
      color: var(--GrayPrimary--translucent);
    }

    &:active {
      border-color: var(--White);
      background-color: var(--White);
      color: var(--GrayPrimary--translucent);
    }
  }

  &[disabled] {
    border-color: var(--Transparent--white);
    background-color: var(--GrayStrokeAndDisabled--translucent);
    color: var(--White);
  }
}

// Purposely very specific name WhiteCTA as this is currently
// only used on the CMS hero for the CTA 'Check my price' button
.Button.WhiteCTA {
  background: var(--White);
  padding: 16px var(--Space-32);
  @include for-phone-only {
    width: 100%;
    max-width: 400px;
  }
  @include for-tablet-and-up {
    width: auto;
  }
  &:hover,
  &:active {
    background: var(--GrayStrokeAndDisabled--opaque);
  }
}

.Button.Stateful {
  &.Medium {
    min-width: var(--ButtonMediumStatefulMinWidth);
  }

  &.White {
    background-color: var(--White);
  }

  @mixin button-stateful-is-active-or-selected {
    // These are used in both the HTML "active" state (mid-click), as well as
    // the custom "is selected" state (this button is in the "on" state).
    border-color: var(--BrandForest);
    background-color: var(--BrandForest);
    color: var(--White);
  }

  border-color: var(--GrayStrokeAndDisabled--translucent);
  background-color: var(--Transparent--white);
  color: var(--GrayPrimary--translucent);

  &:not([disabled]) {
    &:hover {
      border-color: var(--GrayStrokeAndDisabled--translucent);
      background-color: var(--GrayLightHover--translucent);
      color: var(--GrayPrimary--translucent);
    }

    // Bugfix: Need this one to have more specificity then ^ hover
    &.isSelected {
      @include button-stateful-is-active-or-selected;
    }

    &:active {
      @include button-stateful-is-active-or-selected;
    }
  }

  &[disabled] {
    border-color: var(--Transparent--white);
    background-color: var(--GrayStrokeAndDisabled--translucent);
    color: var(--White);
  }

  // This class will be applied with the button is in the "selected" state, e.g.
  // after the user clicks it, or if they have previously selected this value.
  // Note thatthese styles should appear even if the button is disabled, though
  // this probably only makes sense for `.isSelected` (not `:active`).
  &.isSelected {
    @include button-stateful-is-active-or-selected;
  }
}
