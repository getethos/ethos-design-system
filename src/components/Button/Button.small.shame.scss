// ************************************************************** //
// PLEASE ONLY USE IN UniversalNavbar!                            //
// Currently a one-off only for UniversalNavbar                   //
// ************************************************************** //
@import '../Media/Media.scss';
@import '../Spacer/Spacer.scss';

// Mobile first so this starts as the smallest
.Button.Small {
  font-size: calc(var(--compact-font-size) * 1px);
  line-height: 18px;
  padding: 9.5px var(--Space-16) 6.5px;
}

// Reponsive is for the one-off use case of header. Our regular
// Button.Medium sizes do NOT change at time of writing.
@include for-tablet-only {
  /* Same as mobile */
  .Button.Small {
    margin-right: 0;
  }
}
@include for-laptop-only {
  /* Very slightly different */
  .Button.Small {
    padding: 11.5px var(--Space-16) 8.5px;
    font-size: 15.5px;
    line-height: 20px;
    margin-right: 0;
  }
}
@include for-desktop-only {
  /* Different */
  .Button.Small {
    padding: 11.5px 18px 8.5px;
    font-size: 15.5px;
    line-height: 20px;
    margin-right: 0;
  }
}
