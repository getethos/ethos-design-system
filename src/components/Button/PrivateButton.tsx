import * as PropTypes from 'prop-types'
import * as React from 'react'
import useIncludes from '../../hooks/useIncludes.js'
import useInvalid from '../../hooks/useInvalid.js'
import { ArrowIconInline } from './ArrowIconInline'
import styles from './Button.module.scss'
type PrivateButtonProps = {
  ariaLabelId?: string
  role?: string
  'data-tid'?: string
  disabled?: boolean
  fullWidth?: boolean
  name?: string
  onClick?: (...args: any[]) => any
  isSelected?: boolean
  type?: any
  arrowIcon?: boolean
  backArrowIcon?: boolean
  size?: any
  style?: any
}
const HTML_TYPES = { BUTTON: 'button', SUBMIT: 'submit' } // read the docs^
const PUBLIC_PROPS = {
  ariaLabelId: PropTypes.string,
  role: PropTypes.string,
  children: PropTypes.node,
  'data-tid': PropTypes.string,
  disabled: PropTypes.bool,
  fullWidth: PropTypes.bool,
  name: PropTypes.string,
  onClick: PropTypes.func,
  isSelected: PropTypes.bool,
  type: PropTypes.oneOf(Object.values(HTML_TYPES)),
  arrowIcon: PropTypes.bool,
  backArrowIcon: PropTypes.bool,
}
const SIZES = {
  // ***** ONLY USE SMALL IN UniversalNavbar! ***** //
  SMALL: 'Small',
  MEDIUM: 'Medium',
  UNSIZED: 'Unsized',
}
const STYLES = {
  BLACK: 'Black',
  BLACK_OUTLINE: 'BlackOutline',
  // WHITE: 'White', // TODO pending spec
  WHITE_OUTLINE: 'WhiteOutline',
  STATEFUL: 'Stateful',
  STATEFUL_WHITE: 'Stateful White',
  // For semantic <buttons> that are not styled as buttons:
  UNSTYLED: 'Unstyled',
  // Only used for CMS 'Check my price' CTA button on hero
  WHITE_CTA: 'WhiteCTA',
}
export const PrivateButton: React.SFC<PrivateButtonProps> & {
  HTML_TYPES: typeof HTML_TYPES
  PUBLIC_PROPS: typeof PUBLIC_PROPS
  SIZES: typeof SIZES
  STYLES: typeof STYLES
} = ({
  children,
  disabled,
  fullWidth,
  role,
  ariaLabelId,
  name,
  type = PrivateButton.HTML_TYPES.BUTTON,
  size,
  style,
  onClick,
  isSelected, // only used for Button.*.Stateful, aka SelectableHtmlButtonGroup
  arrowIcon,
  backArrowIcon,
  ...rest
}) => {
  // Verify that size, type, and style were valid enum values
  const [isLegalSize] = useIncludes(PrivateButton.SIZES)
  isLegalSize(size)
  const [isLegalType] = useIncludes(PrivateButton.HTML_TYPES)
  isLegalType(type)
  const [isLegalStyle] = useIncludes(PrivateButton.STYLES)
  isLegalStyle(style)
  // Verify that other required props were supplied
  const [, includesKeysOrThrow] = useIncludes(['data-tid'])
  includesKeysOrThrow(rest)
  // Verify that no invalid props were supplied
  const [includesInvalid] = useInvalid(Object.keys(PrivateButton.PUBLIC_PROPS))
  includesInvalid(rest)
  // Generate list of css classes
  // style may be 'Stateful White' so we need to split that
  const classNames = ['Button', size].concat(style.split(' '))
  if (fullWidth) classNames.push('fullWidth')
  if (isSelected) classNames.push('isSelected')
  if (arrowIcon || backArrowIcon) classNames.push('arrowIcon')
  let checked
  if (isSelected) {
    classNames.push('isSelected')
    checked = { 'aria-checked': isSelected }
  }
  const cssModulesClasses = classNames.reduce(
    (accumulator, klass, i) =>
      accumulator.concat(
        `${styles[klass]}${i < classNames.length - 1 ? ' ' : ''}`
      ),
    ''
  )
  return (
    <button
      {...checked}
      className={cssModulesClasses}
      disabled={disabled}
      type={type}
      name={name}
      onClick={onClick}
      data-tid={rest['data-tid']}
      role={role}
    >
      {backArrowIcon && <ArrowIconInline shouldFlip />}
      {ariaLabelId ? <span id={ariaLabelId}>{children}</span> : children}
      {arrowIcon && <ArrowIconInline />}
    </button>
  )
}
PrivateButton.HTML_TYPES = HTML_TYPES
PrivateButton.PUBLIC_PROPS = PUBLIC_PROPS
PrivateButton.SIZES = SIZES
PrivateButton.STYLES = STYLES
