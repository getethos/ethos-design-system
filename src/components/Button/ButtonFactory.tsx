import * as React from 'react'
import { PrivateButton } from './PrivateButton'
export function ButtonFactory(privateProps: any) {
  const PublicButtonComponent = (downstreamProps: any) => {
    return <PrivateButton {...downstreamProps} {...privateProps} />
  }
  return PublicButtonComponent
}
