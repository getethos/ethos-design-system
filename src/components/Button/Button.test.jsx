import React from 'react'
import { Button } from './Button'
import renderer from 'react-test-renderer'

describe('Button', () => {
  describe('API', () => {
    test('has correct shape', () => {
      expect(Button).toBeDefined()
      expect(Button.Medium).toBeDefined()
      expect(Button.Medium.Black).toBeDefined()
      expect(Button.Medium.BlackOutline).toBeDefined()
      expect(Button.Medium.WhiteOutline).toBeDefined()
      expect(Button.Medium.Stateful).toBeDefined()
      expect(Button.Medium.Stateful.Default).toBeDefined()
      expect(Button.Medium.Stateful.White).toBeDefined()
      expect(Button.Unstyled).toBeDefined()
    })
  })

  describe('default rendering', () => {
    test('Button.Medium.Black', () => {
      const tree = renderer
        .create(
          <Button.Medium.Black data-tid="foo" name="bar" disabled={false}>
            test
          </Button.Medium.Black>
        )
        .toJSON()
      expect(tree).toMatchSnapshot()
    })
    test('Stateful isSelected', () => {
      const tree = renderer
        .create(
          <Button.Medium.Stateful.Default
            isSelected={true}
            data-tid="foo"
            name="bar"
          >
            test
          </Button.Medium.Stateful.Default>
        )
        .toJSON()
      expect(tree).toMatchSnapshot()
    })
    test('Stateful isSelected White', () => {
      const tree = renderer
        .create(
          <Button.Medium.Stateful.White
            isSelected={true}
            data-tid="foo"
            name="bar"
          >
            test
          </Button.Medium.Stateful.White>
        )
        .toJSON()
      expect(tree).toMatchSnapshot()
    })
  })
})
