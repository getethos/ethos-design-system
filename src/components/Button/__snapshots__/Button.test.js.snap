// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button default rendering Array<ReactNode> 1`] = `
<button
  className="Button Medium Black"
  data-tid="foo"
  data-track-button={true}
  disabled={false}
  name="bar"
  type="button"
>
  <div />
  <div />
  3
  string
  Fragment with key
</button>
`;

exports[`Button default rendering Button.Medium.Black 1`] = `
<button
  className="Button Medium Black"
  data-tid="foo"
  data-track-button={true}
  disabled={false}
  name="bar"
  type="button"
>
  test
</button>
`;

exports[`Button default rendering Button.Medium.Salamander 1`] = `
<button
  className="Button Medium Salamander"
  data-tid="foo"
  data-track-button={true}
  disabled={false}
  name="bar"
  type="button"
>
  test
</button>
`;

exports[`Button default rendering Stateful isSelected 1`] = `
<button
  aria-checked={true}
  className="Button Medium Stateful isSelected isSelected"
  data-tid="foo"
  data-track-button={true}
  disabled={false}
  name="bar"
  type="button"
>
  test
</button>
`;

exports[`Button default rendering Stateful isSelected White 1`] = `
<button
  aria-checked={true}
  className="Button Medium Stateful White isSelected isSelected"
  data-tid="foo"
  data-track-button={true}
  disabled={false}
  name="bar"
  type="button"
>
  test
</button>
`;
